File: zig/isogram/Iteration2/isogram.zig
pub fn isIsogram(str: []const u8) bool {

    var characterFlags: u32 = 0;

    for (str) |character| {
        if (charToInt(character)) |charIndex| {
            const mask = @as(u32, 1) << charIndex;

            if (characterFlags & mask != 0) return false;
            characterFlags |= mask;
        }
    }

    return true;
}

fn charToInt(c: u8) ?u5 {
    return switch (c) {
        'a'...'z' => @intCast(c - 'a'),
        'A'...'Z' => @intCast(c - 'A'),
        else => null,
    };
}

File: c/isogram/Iteration2/isogram.c

#include <stddef.h>

#include "isogram.h"

bool is_isogram(const char phrase[])
{
    if (phrase == NULL) return false;
    if (*phrase == '\0') return true;
    
    unsigned int characterCounts[26] = { 0 };
    
    for (const char* c = phrase; *c != '\0'; c++)
    {
        char character = *c;
        
        if (character >= 'a' && character <= 'z')
        {
            int charIndex = character - 97;
            characterCounts[charIndex]++;
        }
    
        if (character >= 'A' && character <= 'Z')
        {
            int charIndex = character - 65;
            characterCounts[charIndex]++;
        }
    }
    
    for (int i = 0; i < 26; i++)
    {
        if (characterCounts[i] > 1) return false;
    }
    
    return true;
}